---
layout: tutorial_slides
logo: "GTN"

title: "Genome assembly and assembly QC - Introduction short version"
questions:
  - "What do I need to do before starting a genome assembly project?"
  - "Quick overview of the steps before the bioinformatic"
  - "Definitions of bioinformatics terms for assembly"
  - "Definitions of tools used to assess the quality of an assembly"
key_points:
  - "We learned the importance of preparing the project to ensure its success"
  - "We learned the importance of surrounding ourselves with all the people who have knowledge of the different parts of the project (wet lab, sequencing, bioinformatics,...)"
  - "We learned the definitions of bioinformatics terms used in genomes assembly"
  - "We have seen the bioinformatics file formats used for these analyses"
To go further : [Deeper look into Genome Assembly algorithms](https://training.galaxyproject.org/training-material/topics/assembly/tutorials/algorithms-introduction/slides.html#1)
  - "We have seen the bioinformatics tools to assess the quality of an assembly"
To go further : [Genome assembly quality control](https://training.galaxyproject.org/training-material/topics/assembly/tutorials/assembly-quality-control/slides.html#1)
contributions:
  authorship:
  - scorreard
  - alexcorm
  - lleroi
  editing:
  - abretaud
  - stephanierobin
  - r1corre
  funding:
  - gallantries
level: Introductory
time_estimation: "1h"


---
### Steps before starting a genome project

.left[
<br>
- **Step 1**: Gather information about the target species : location, overlap with Indigenous People or local communities territories, expected genome size, ploidy, micro-chromosomes, organelles, etc.
What type of data is already available (reference gneomes, raw data, etc) : INSDC
<br>
- **Step 2**: If appropriate (depending on species and sampling location), engage with local communities, Indigenous people and/or local scientists who may have Traditional Ecological Knowledge (TEK) associated to the species.
Obtain appropriate consent and permits (Nagoya protocol) before sampling.
<br>
- **Step 3**: Build a broad community of collaborators for the project, if possible.
<br>
- **Step 4**: Select the best possible DNA source and an optimal extraction procedure - **Sampling is THE key step**
<br>
- **Step 5**: Choose an appropriate sequencing technology
<br>
- **Step 6**: Sequence and assemble!
]

---

### Steps before starting a genome project - ERGA model

.image-1[
  ![ERGA process](../../images/ERGA_process.jpg)
]

.footnote[https://www.nature.com/articles/s44185-024-00054-6]

---

### Genome information: Genome availability, expected genome size, ploidy, etc

.pull-left[
**How to collect informations?**

- [GoaT](https://goat.genomehubs.org/) (Genome on a Tree)
- Bibliography

.image-1[
  ![GOAT logo](../../images/goat_logo.png)
]
]

.pull-right[
.image-100[
  ![variation in estimated genome sizes in base pairs](../../images/genome_assembly_introduction/genomes-size.png)
]]

.footnote[https://commons.wikimedia.org/w/index.php?curid=19537795]


---

### Genome information: Ploidy level

.pull-right[
.image-55[
  ![Heterozigous genotype](../../images/genome_assembly_introduction/ploidy.png)
]]

.pull-left[

**Ploidy (N):**

Number of sets of chromosomes in a cell


| Organism                  | Ploidy    |
|:--------------------------|:---------:|
| Bacteria                  | 1N        |
| Human, mouse, rat         | 2N        |
| Amphibians (*Xenopus*)    | 2N to 12N |
| Plants (wheat)            | 6N        |
| **Autopolyploid**         | .         |
| **Hybrids**               | .         |

]

**Higher ploidy -> harder to assemble => Increase of sequencing depth**

.footnote[Daniel Hartl. Essential Genetics: A Genomics Perspective. Jones & Bartlett Learning. p. 177. ISBN 978-0-7637-7364-9. (2011).]

---

### Genome information: Heterozygosity level

.pull-left[
.left[
**Heterozygous:** Locus-specific,  diploid (2N) organism has two different alleles of a particular gene at the same locus

<br>
**Heterozygosity** is a metric used to indicate the probability that an individual is heterozygous for a particular allele

]]

.pull-right[
.image-100[
  ![Heterozigous genotype](../../images/genome_assembly_introduction/heterozygous.jpg)
]]

**Higher heterozygosity -> harder to assemble => Increase of sequencing depth**

.footnote[https://www.genome.gov/genetics-glossary/heterozygous]

---

### Genome information: Others

- **Karyotype:** chromosome number
<br>
- **Sex chromosome system:** None, XY, ZW, UV,…
<br>
- **Purity:** possible presence of contaminants and/or symbionts?
<br>
- Is there any other **useful data (NCBI, SRA, ENA, etc)** that could improve my assembly?

---

### DNA extraction tips

.left[
- Many **DNA extraction protocol** are available for a wide range of species/taxa (VGP, Darwin Tree of Life, Nanopore, PacBio, etc)
<br>
- **Keep DNA samples** from the same individual in case of library preparation or sequencing failure, need more coverage, new sequencing technology, etc
<br>
- Use a **single individual** and sequence a **haploid, a highly inbred diploid** organism, or an **isogenic** individual

]

---

### Sequencing steps - The options

.left[
This mainly **depends on the quantity and quality of DNA as well as the cost of the experiment** but many parameters need to be considered before performing an NGS experiment:

- Short **versus** long reads or **both**
- Read length
- Read quality/error rate
- Genome read coverage/depth : Number of unique reads that include a given nucleotide in the reconstructed sequence.
    -30X coverage mean that, on average, each nucleotide in the genome is covered bu 30 reads.
    - Coverage = (read count * read length ) / total genome size
- Library preparation
- Available technology
- Downstream applications
]

---

### Sequencing steps - The technologies

.left[
Sequencing technology for **assembly**:
- **PacBio Hifi**: long reads (up to 20kb)
- **Nanopore**: long reads and ultra-long read (up to 100kb)
- **Illumina or MGI**: short reads (up to 2x250bp) with high quality reads. Sequencing bias with AT/GC rich regions

------

Sequencing technology for **scaffolding**:
- **Hi-C:** **restriction enzyme fragmentation** (single, multiples sites or DNAse). Need huge amount of coverage.
Providers : Arima Genomics, Phase Genomics, Dovetail Genomics
- Optical mapping: technique to physically locate specific enzymes restriction sites or sequence motifs to produce DNA sequence fingerprints.

Providers : BioNano, BGI
- Mate pair (deprecated)
- BAC/YAC/Fosmids (deprecated)

------

Typical sequencing strategies - EBP (Earth Biogenome Project) recommendation -
- Long-reads (PacBio HiFi or ONT) : 15x per haplotype
- Hi-C data (Arima / Illumina)
Polishing is no longer necessary or recommended

]


---

### Bioinformatics steps - Defintions

.image-60[
  ![Illustration of the working principle of scaffolding](../../images/Schematic-representation-of-the-method-used-to-assemble-Illumina-reads-into-contigs-and.png)
]

.left[
**Contig**: a contiguous sequence in an assembly. A contig does not contain long stretches of unknown sequences (aka assembly **gaps**).
The contig is usually generated using the long-reads data.
<br>
**Scaffold**: a sequence consists of one or multiple contigs connected by assembly gaps of typically inexact sizes. A scaffold is also called a **supercontig**, though this terminology is rarely used nowadays.
Usually, scaffolds are generated using the Hi-C data
<br>
**Assembly**: a set of contigs or scaffolds.
]

.footnote[Johnson, Marc & Carpenter, Eric & Zhijian et al. (2012). Evaluating Methods for Isolating Total RNA and Predicting the Success of Sequencing Phylogenetically Diverse Plant Transcriptomes. PloS one. 7. e50226. 10.1371/journal.pone.0050226.]

---

### Bioinformatics steps - Assembly quality

.left[
Different level of assembly exist today :
<br>
- **Contig Assembly**
<br>
- **Scaffold assembly**
<br>
- **Chromosome-level assembly** : When the number of scaffolds is the number of expected chromosomes, it means that 1 chromosome = 1 scaffold, and no large string of sequence is unlocated.

It can still contain gaps in between the scaffolds (shown as "NNNNNNNN" in the assembly)
<br>
- **T2T (Telomere-to-Telomere)** : Assembly without any gap (a chromosome level assembly without "NNNNN" sequences) 
]

---

### Bioinformatics steps - Defintions

.left[
**Haplotig**: a contig that comes from the same haplotype. In an unphased assembly, a contig may join alleles from different parental haplotypes in a diploid or polyploid genome.
<br>
**Primary assembly**: a complete assembly with long stretches of phased blocks.
<br>
**Alternate assembly**: an incomplete assembly consisting of haplotigs in heterozygous regions. An alternate assembly always accompanies a primary assembly. It is not useful by itself as it is fragmented and incomplete.
<br>
**Haplotype-resolved assembly**: sets of complete assemblies consisting of haplotigs, representing an entire diploid/polyploid genome.
]

.image-60[
  ![Illustration of the assembly types](../../images/genome_assembly_introduction/assemblies_definitions.png)
]

---

### Computational resources and requirements

.left[
To be successful, you must have **sufficient computing resources (CPUS, RAM, walltime and storage)**.
- The resources needed are **different** for each step:
  - Assembly
  - Annotation
  - Other analysis tools
  <br>
- For genome **assembly:**
  - Running times and RAM increase with data **type** and **amount**
  - More data for large genomes, increase **runtime/RAM/Storage**
  - Most of tools **run** on a **single node**: they are parallelized but not distributed
  <br>
- For genome **annotation:**
  - Mapping/alignment of external data (RNA-seq, proteins) can be parallelized and distributed
  - Annotation process can be parallelized and distributed
]


---

### Bioinformatics data formats

.left[
**FASTA**: a text-based format for representing either nucleotide sequences or amino acid (protein) sequences, in which nucleotides or amino acids are represented using single-letter codes.
]

.image-100[
  ![Fasta format description](../../images/genome_assembly_introduction/fasta_format.png)
]
Image licensed CC-BY 4.0 {% cite datacompression2016 %}

.footnote[Hosseini, M., Pratas, D. & Pinho, A. J. A Survey on Data Compression Methods for Biological Sequences. Information 7, 56 (2016).]

---

### Bioinformatics data formats

.left[
**FASTQ**: a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores (Phred). Both the sequence letter and quality score are each encoded with a single ASCII character for brevity. It's the standard sequencing output for Illumina and MGI sequencers.
]

.image-100[
  ![Fastq format description](../../images/genome_assembly_introduction/fastq_format.png)
]
Image licensed CC-BY 4.0 {% cite datacompression2016 %}

.footnote[Hosseini, M., Pratas, D. & Pinho, A. J. A Survey on Data Compression Methods for Biological Sequences. Information 7, 56 (2016).]

---

### Bioinformatics data formats

**SAM (Sequence Alignment Map)**: a text-based format originally for storing biological sequences aligned to a reference sequence developed by Heng Li and Bob Handsaker et al.
<br>
**BAM (Binary Alignment Map)**: the comprehensive raw data of genome sequencing; it consists of the lossless, compressed binary representation of the SAM format. It's the standard sequencing output for PacBio sequencers.
<br>
**CRAM (Compressed Reference-oriented Alignment Map)**: a compressed columnar file format for storing biological sequences aligned to a reference sequence.

.pull-left[
.image[
  ![SAM format description](../../images/genome_assembly_introduction/sam_format.png)
]]

.pull-right[
Image licensed CC-BY 4.0 {% cite datacompression2016 %}
]

.footnote[Hosseini, M., Pratas, D. & Pinho, A. J. A Survey on Data Compression Methods for Biological Sequences. Information 7, 56 (2016).]

---

### After the assembly, how do we assess its quality?

.image-100[
  ![The 3C for genome assembly quality control](../../images/genome-qc/3C.png)
]

---

### Continuity : N50

.left[
**N50**: given a set of sequences of varying lengths, the N50 is defined as **the length L of the shortest contig** for which **longer and equal length contigs cover at least 50% of the assembly**.
<br>
**L50**: given a set of sequences of varying lengths, the L50 is defined as **count of smallest number of sequences** whose **length sum makes up 50% of the assembly**.
<br>
**N50 describes a sequence length whereas L50 describes a number of sequences.**

Example:
- Genome size = 100
- Sequence sorted by size list L = (25, 10, 10, 8 , 7, 7 , 6 , 5, 5, 5, 5, 3, 2, 2 ) = 100
- 50% of the total length is contained within sequences of at least 8bp: 25 + 10 + 10 + 8 ≥ 50
]

.image-100[
  ![Schematic explanation of N50](../../images/genome-qc/N50.png)
]

**N50 = 8** and **L50 = 4**

.footnote[Alhakami, H., Mirebrahim, H., & Lonardi, S. (2017). A comparative evaluation of genome assembly reconciliation tools. Genome biology, 18(1), 1-14.]

---

.pull-right[
.image-55[
  ![Quast report](../../images/genome-qc/Quast_report.jpg)
]]

.pull-left[
### Tool to evaluate continuity : QUAST

- **QUAST**: for genome assemblies.
- **MetaQUAST**: for metagenomic datasets.
- **QUAST-LG**: for large genomes (e.g., mammalians).
- **rnaQUAST**: for RNAseq.
- **Icarus**: an interactive visualizer for these tools.
]

---

### Completeness : BUSCO score

.pull-left[
**BUSCO**: Assessing genome assembly and annotation completeness with **B**enchmarking **U**niversal **S**ingle-**C**opy **O**rthologs
<br>
Quantitative assessment of genome assembly based on evolutionarily informed expectations of gene content from near-universal single-copy orthologs.
.image-70[
![Formula to estimate assembly completeness for core genes](../../images/genome-qc/busco-formula.png)
]
]
.pull-right[
.image-70[
![Example of BUSCO plot for Nosema species (Microsporidia) ](../../images/genome-qc/BUSCO-scores.png)
]]

.footnote[Tips: Reference databases are constructed using known genomes. Species with few/no close genomes available can have very bad scores.]

---

### Correctness

** Proportion of the assembly that is free from mistakes**
<br>
<br>
- Indels / SNPs
- Mis-joins
- Repeat compressions
- Unnecessary duplications
- Rearrangements
<br>
<br>
**→ Align back reads to the assembly and check for inconsistencies**

---

### Evaluation against reference genome (or second haplotype)

.image-30[
![Example of a dot plot between 2 genomes.](../../images/genome-qc/dotplot-dgenies.png)
]

---

### Assembly QC Tips

- The quality of an assembly is often validated by using other data from the same individual or from other individuals (RNA-Seq alignment, Hi-C alignment, DNA-Seq alignment,...).

- The positions of the telomeric repeats in the chromosome assemblies are also of interesting to evaluate the correctness.

- The identification of organelles (mitochondria, chloroplast,...) can also inform us about the quality of the assembly in terms of completness. However, the structure of the organelles may lead the assembler to think that they are repeats and he discards them.

- In the case of diploid organisms, one of the classical problems of assemblies is the conservation of the two haplotypes. We obtains particular BUSCO / kmer / assembly size metrics that can be corrected by removing, "purging", the haplotigs.


